cmake/BuildBPF.cmake

```diff
@@ -0,0 +1,153 @@
+  REQUIRED)
+
+function(btf_header NAME)
+  cmake_parse_arguments(
+    ARG
+    ""
+    "OUTPUT;SOURCE"
+    ""
+    ${ARGN}
+  )
+  if (NOT DEFINED ARG_SOURCE)
+    set(ARG_SOURCE "/sys/kernel/btf/vmlinux")
+  endif ()
+  if (NOT DEFINED ARG_OUTPUT)
+    set(ARG_OUTPUT "${NAME}.h")
+  endif ()
+  add_custom_command(
+    OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/${ARG_OUTPUT}
+    COMMAND ${BPFTOOL} btf dump file "${ARG_SOURCE}" format c > ${CMAKE_CURRENT_BINARY_DIR}/${ARG_OUTPUT}
+    VERBATIM
+  )
+  add_custom_target(${NAME}
+    DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/${ARG_OUTPUT}
+  )
+endfunction()
+
+function(bpf NAME)
+  cmake_parse_arguments(
+    ARG
+    ""
+    "BITCODE;OBJECT;BINARY;BTF;SOURCE;BTF_BASE;FUNCTIONS"
+    "DEPENDS"
+    ${ARGN}
+  )
+  if (NOT DEFINED ARG_SOURCE)
+    set(ARG_SOURCE "${NAME}.c")
+  endif ()
+  if (NOT DEFINED ARG_OBJECT)
+    set(ARG_OBJECT "${NAME}.o")
+  endif ()
+  # The joys of BPF & debug tooling 101.
+  #
+  # The purpose for this function is to generate various types of output for a
+  # given source file, for testing and embedding purposes:
+  #  * LLVM bitcode file compatible with our major version of LLVM.
+  #  * A binary annotated with DWARF type information.
+  #  * Encoded BTF type information for this source unit.
+  #  * A list of functions defined in the file.
+  #
+  # Naturally, you'd assume that this could be implementated mostly as a chain
+  # of steps, save the intermediate results along the way. Wrong! Unfortunately,
+  # each of these steps has their own happy path.
+  #
+  # First, the bitcode. This is actually straight-forward! We need to ensure
+  # that this gets built with the correct version of LLVM (not whatever
+  # compiler CMAKE has selected) so that it is compatible with the LLVM
+  # libraries that we are linking against. We can even correctly select bpf as
+  # our target, to get the extent that there is any early specialization.
+  #
+  # Next, you might assume that taking this bitcode and turning it into a BPF
+  # object is the logical thing to do, in order to ensure that all the types
+  # are compatible and consistent. In fact, this even *seems* to work. But
+  # after banging your head on the keyboard for a while, you'll realize that
+  # something is broken. As you contort clang, and try building from scratch,
+  # you'll realize that clang seems to generate proper annotations and
+  # relocations for loadable programs (the happy path for that!), but fails to
+  # generate the same types for base types. In fact, in doesn't even generate
+  # named arguments without an explicit `-O2`.  Oh well, I suppose that the
+  # common path for the kernel is to build with `gcc` and then rely on the
+  # `pahole` conversion. Let's do that.
+  #
+  # But hold on a second, we said that we wanted to build a BPF target? Turns
+  # out that `gcc` can't do that very well. So `clang` can't build our BPF
+  # target because of incorrect BTF, and `gcc` can't build the BPF target. So
```

Comment:
Did you try with these magic flags? https://github.com/libbpf/libbpf-rs/blob/f9ec2c444032454c2ca791c8058fc3499ee51818/libbpf-cargo/src/build.rs#L66

Building with clang should work...

Reply:
Yeah, the building was not the problem. It just had some weird issues with the per-CPU attributes on the BTF. I'm happy to rip this out and revisit it, but for right now I'm fine with it working as-is (since it's how it works currently with your version).

It could have been the case that it was a bug with `clang-18` (that I was testing with here), because the change pins that version (since it is necessary for the bitcode). But this is actually an argument in favor of just leaving this as `gcc`.